File: AlgorithmPractice/150/HashMaps/242_ValidAnagram.py
class Solution(object):
    def isAnagram(self, s, t):
        if len(s) != len(t):
            return False
 
        countS, countT = {}, {}

        for i in range(len(s)):
            countS[s[i]] = 1 + countS.get(s[i], 0)
            countT[t[i]] = 1 + countT.get(t[i], 0)
            
        return countS == countT
    
    # O(n) time, both strings are the same length (n).
    # O(1) memory, if the character set is fixed,
    # worst case with a large or variable character set, it could be O(n) space.

File: AlgorithmPractice/150/HashMaps/242_ValidAnagram.ts
function isAnagram(s: string, t: string): boolean {
    if (s.length !== t.length) return false;

    const countS = new Map<string, number>();
    const countT = new Map<string, number>();

    for (let i = 0; i < s.length; i++) {
        const charS = s[i];
        const charT = t[i];

        countS.set(charS, (countS.get(charS) || 0) + 1);
        countT.set(charT, (countT.get(charT) || 0) + 1);
    }

    for (let [key, value] of countS) {
        if (countT.get(key) !== value) {
            return false;
        }
    }

    return true;
};